### PR TITLE
Framework add/update by yaml

### PR DESCRIPTION
**Pull request by stinkyfingers**
# Description
https://shipaio.atlassian.net/secure/RapidBoard.jspa?rapidView=1&projectKey=SHIPA&modal=detail&selectedIssue=SHIPA-1546

Adds ability to add & update frameworks by yamll file, e.g. `ketch add framework myframework.yaml` and `ketch add framework ./mydir/myframework.yaml`.

Notes:
- If the user specifies a .yaml file as the arg to `framework add` or `framework update`, additional flags are ignored.
- During `framework add` ingress `ClassName` and `IngressType` are defaulted to `traefik` if the yaml files's fields are empty, similar to the flag-based approach. Fields `Version`, `AppQuotaLimit`, and `NamespaceName` also are defaulted if a value is not provided.
- I switched the `FrameworkSpec.AppQuotaLimit` field to a `*int`, so that we could determine during yaml unmarshaling if the field was zero or nil. The default is -1, so unmarshaling to a `0` when the field is unpopulated won't work. This change required dereferencing appQuotaLimit in a lot of places.

Fixes # 1546

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Chore (documentation addition or typo, file relocation)

## Testing

- [x] New tests were added with this PR that prove my fix is effective or that my feature works (describe below this bullet)
- [ ] This change requires no testing (i.e. documentation update)

## Documentation

- [x] All added public packages, funcs, and types have been documented with doc comments
- [x] I have commented my code, particularly in hard-to-understand areas

## Final Checklist:

- [x] I followed standard [GitHub flow](https://guides.github.com/introduction/flow/) guidelines
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings

## Additional Information (omit if empty)

Please include anything else relevant to this PR that may be useful to know.

### CODE REVIEW

File: cmd/ketch/framework_update_test.go [hunk 1]
```diff
@@ -44,20 +60,68 @@ func Test_frameworkUpdate(t *testing.T) {
...
-				CtrlClientObjects: []runtime.Object{frontendFramework},
+				CtrlClientObjects:    []runtime.Object{frontendFramework},
+				DynamicClientObjects: []runtime.Object{clusterIssuerStaging},
+			},
+			options: frameworkUpdateOptions{
+				name:          file.Name(),
+				appQuotaLimit: 10,
+			},
+			before: func() {
+				file.Truncate(0)
```
Comment (aleksej-paschenko) [inline cmd/ketch/framework_update_test.go:92]: looks tricky
how about `before()` returning a filename and we use it to set `options.name`. 

my point is that having functions with side effects creates additional complexity
```golang
yamlFilename := tt.createFrameworkYamlFile()
options.name = yamlFilename

err := frameworkUpdate(context.Background(), tt.cfg, tt.options, out)

```
vs
```golang
	if tt.before != nil {
		tt.before()
	}
	out := &bytes.Buffer{}
	err := frameworkUpdate(context.Background(), tt.cfg, tt.options, out)
```
something is going on there but I can't see it)

File: cmd/ketch/framework_update_test.go [hunk 2]
```diff
@@ -44,20 +60,68 @@ func Test_frameworkUpdate(t *testing.T) {
 			},
 		},
 	}
+	file, err := ioutil.TempFile("", "*.yaml")
```
Comment (aleksej-paschenko) [inline cmd/ketch/framework_update_test.go:63]: how about  `t.TempDir`?

File: cmd/ketch/framework_add_test.go [hunk 3]
```diff
@@ -30,17 +33,88 @@ func Test_addFramework(t *testing.T) {
 			},
 		},
 	}
+	file, err := os.CreateTemp("", "*.yaml")
```
Comment (aleksej-paschenko) [inline cmd/ketch/framework_add_test.go:36]: t.TempDir()?

File: cmd/ketch/framework_add_test.go [hunk 4]
```diff
@@ -129,23 +202,25 @@ func Test_addFramework(t *testing.T) {
...
 			wantErr: ErrClusterIssuerNotFound.Error(),
 		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
+			if tt.before != nil {
+				tt.before()
+			}
 			out := &bytes.Buffer{}
 			err := addFramework(context.Background(), tt.cfg, tt.options, out)
```
Comment (aleksej-paschenko) [inline cmd/ketch/framework_add_test.go:214]: check a comment below about functions with side-effects please

File: cmd/ketch/framework_add.go [hunk 5]
```diff
@@ -62,7 +61,8 @@ func newFrameworkAddCmd(cfg config, out io.Writer, addFramework addFrameworkFn)
 }
 
 type frameworkAddOptions struct {
-	name string
+	version string
+	name    string
```
Comment (aleksej-paschenko) [inline cmd/ketch/framework_add.go:65]: can we have a comment that `name` can be either a framework name or a file name?

**Review by aleksej-paschenko — CHANGES_REQUESTED** [review]

File: cmd/ketch/framework_update_test.go (hunk 1 above)
Comment (stinkyfingers) [reply cmd/ketch/framework_update_test.go]: Good call. I removed the `before()` func from all these new tests. I opted to create a temp file and write test data on each test loop. There is a little duplicated code in the `t.Run()` functions, but it's nice to be explicit about what is happening in tests.

File: cmd/ketch/framework_update_test.go (hunk 2 above)
Comment (stinkyfingers) [reply cmd/ketch/framework_update_test.go]: 👍

File: cmd/ketch/framework_add_test.go (hunk 3 above)
Comment (stinkyfingers) [reply cmd/ketch/framework_add_test.go]: 👍

File: cmd/ketch/framework_add_test.go (hunk 4 above)
Comment (stinkyfingers) [reply cmd/ketch/framework_add_test.go:198]: 👍

File: cmd/ketch/framework_add.go (hunk 5 above)
Comment (stinkyfingers) [reply cmd/ketch/framework_add.go]: Comment added. I was worried this would be a little unclear, but I don't have a better solution in mind.

**Comment by stinkyfingers** [comment]
Superseded by 112.

**Pull request closed without merging**